_drafts/cortex-m-fault-debug.md

```diff
@@ -0,0 +1,1079 @@
+---
+title: "How to debug an ARM Cortex-M HardFault"
+description: "A step by step guide of how to investigate what led to a fault on a Cortex-M device with examples and strategies about how to recover"
+tag: [cortex-m]
+author: chris
+---
+
+At one point or another when working with an ARM Cortex-M MCU, you will hit something like:
+
+```shell
+(gdb) backtrace
+#0  HardFault_Handler() at startup.c:136
+#1  <signal handler called>
+```
+
+<!-- excerpt start -->
+
+In this article we will demystify how to determine what went wrong when an ARM Cortex-M fault occurs. We will walk step by step through what the registers to inspect and what the values actually mean. We will discuss debug tricks that can be used for some of the most challenging classes of HardFaults, how how to automate the fault analysis, and how to recover from a fault without rebooting the system. Finally, we will apply what we have learned in the article and walk through a couple practical examples.
+
+<!-- excerpt end -->
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest
+posts straight to your mailbox_
+
+## Table of Contents
+
+<!-- prettier-ignore -->
+* auto-gen TOC:
+{:toc}
+
+## Determining What Caused The Fault
+
+All MCUs in the Cortex-M series have several different pieces of state which can be analyzed when a fault takes place to trace down what went wrong.
+
+First we will explore the dedicated fault status registers that are present on all Cortex-M MCUs except the Cortex-M0.
+
+If you are trying to debug a Cortex-M0, you can skip ahead to [this section](#registers-prior-to-exception) where we discuss how to recover the core register state and instruction being executed at the time of the exception.
```

Comment:
Maybe: "you can skip ahead to *the next section*" rather than "this section". "This" implies proximity (i.e. this current one).